### PR TITLE
Content index

### PR DESCRIPTION
**Pull request by gbprz**
#418

First iteration seems to work well, let me know what you think

The solution seemed to be getting too long to leave it in the contents_controller, so I put it in the contents model. Was this a bad move? I can bring everything back to the contents_controller if this is not good style.

### CODE REVIEW

File: app/models/content.rb [hunk 1]
```diff
@@ -179,4 +179,33 @@ def perform_action(action_name, options)
...
+          if params[:type] && content.type != params[:type] then next end
+          filtered_contents.push(content)
+        end
+      end
+    else
+      # If filtering by user or type, we don't need to search through subscriptions, only contents
+      if params[:user] then query_conditions[:user_id] = params[:user] end
+      if params[:type] then query_conditions[:type] = params[:type] end
+      filtered_contents = Content.find(:all, :conditions => query_conditions)
+    end
```
Comment (zr2d2) [inline app/models/content.rb:214]: I think these would be cleaner written as

``` ruby
query_conditions[:feed_id] = params[:feed] if params[:feed]
query_conditions[:screen_id] = params[:screen] if params[:screen]
if query_conditions.empty?
  # code if true
else
  # code if false
end
```

I favor the `if` statements at the end, then you can drop the then and end statements

File: app/views/contents/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<section class="viewblock">
+  <header class="viewblock-header" id="browse-header">
+    <h1 class="default-padding"><%= t('.Concerto Content') %> </h1>
```
Comment (zr2d2) [inline app/views/contents/index.html.erb]: you don't necessarily need the parentheses

Comment (gbprz) [inline app/views/contents/index.html.erb]: Okay, this also reminds me that I need to add the translation

File: app/models/content.rb (hunk 1 above)
Comment (gbprz) [inline app/models/content.rb:214]: Fixed in my latest commits, thanks for the heads up. Is it good to style the if statement before the 'next' the same way? I did so for lines 196 and 197 in my most recent commits

File: app/controllers/contents_controller.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 class ContentsController < ApplicationController
-  before_filter :get_content_const, :only => [:new, :create, :preview]
+  before_filter :get_content_const, :only => [:index, :new, :create, :preview]
```
Comment (bamnet) [inline app/controllers/contents_controller.rb]: I don't think you're using the global that get_content_const defines, are you?

File: config/routes.rb [hunk 4]
```diff
@@ -165,9 +166,10 @@ def matches?(request)
   match 'content/' => 'feeds#index'
   match 'browse/' => 'feeds#index'
 
+
```
Comment (bamnet) [inline config/routes.rb]: Care to remove this extra newline?

File: app/models/content.rb [hunk 5]
```diff
@@ -179,4 +179,39 @@ def perform_action(action_name, options)
   def after_add_callback(unused_submission)
   end
 
+  def self.filter_all_content(params)
+    # Filter Concerto content by user, screen, feed, and/or type
+    filtered_contents = Array.new
+
+    query_conditions = {}
+    # If filtering by screen or feed, we must search through subscriptions
```
Comment (bamnet) [inline app/models/content.rb]: This isn't anywhere close to how I would have approached it (chaining where clauses) but it looks pretty neat, nice work!

File: config/routes.rb (hunk 4 above)
Comment (mfrederickson) [inline config/routes.rb]: lol.  sounds reminiscent of our tab discussion on tuesday.

File: app/models/content.rb [hunk 6]
```diff
@@ -179,4 +179,39 @@ def perform_action(action_name, options)
...
+    if params[:screen] || params[:feed]
+      query_conditions[:feed_id] = params[:feed] if params[:feed]
+      query_conditions[:screen_id] = params[:screen] if params[:screen] 
+      subs = Subscription.find(:all, :conditions => query_conditions)
+      subs.each do |sub|
+        sub.contents.each do |content|
+          # If filtering by user or type, do not add (skip) content that 
+          # does not match filter criteria
+          next if params[:user] && content.user_id.to_s != params[:user]  
+          next if params[:type] && content.type != params[:type]
```
Comment (zr2d2) [inline app/models/content.rb:199]: nice work, I like it!

File: app/models/content.rb [hunk 7]
```diff
@@ -179,4 +179,39 @@ def perform_action(action_name, options)
...
 
+  def self.filter_all_content(params)
+    # Filter Concerto content by user, screen, feed, and/or type
+    filtered_contents = Array.new
+
+    query_conditions = {}
+    # If filtering by screen or feed, we must search through subscriptions 
+    if params[:screen] || params[:feed]
+      query_conditions[:feed_id] = params[:feed] if params[:feed]
+      query_conditions[:screen_id] = params[:screen] if params[:screen]
```
Comment (zr2d2) [inline app/models/content.rb]: I still might put the `query_conditions` outside the `if` statement and use `if !query_conditions.empty?`, or even switch the if/else, and not negate the `.empty?`

Comment (gbprz) [inline app/models/content.rb]: Not sure if I am understanding properly, but replace:

```
if query_conditions.empty?
  filtered_contents = Content.find(:all)
else
  filtered_contents = Content.find(:all, :conditions => query_conditions) 
end 
```

with

```
filtered_contents = Content.find(:all) if query_conditions.empty? 
filtered_contents = Content.find(:all, :conditions => query_conditions) if !query_conditions.empty? 
```

File: config/routes.rb (hunk 4 above)
Comment (gbprz) [inline config/routes.rb]: hahaha, yea ... my bad 
:octocat:

File: app/models/content.rb [hunk 8]
```diff
@@ -179,4 +181,39 @@ def perform_action(action_name, options)
...
+    else
+      # If filtering by user or type, we don't need to search through subscriptions, only contents
+      query_conditions[:user_id] = params[:user] if params[:user]
+      query_conditions[:type] = params[:type] if params[:type]
+      if query_conditions.empty?
+        # No filters are specified
+        filtered_contents = Content.all
+      else
+        # User and/or type filters are specified
+        filtered_contents = Content.all query_conditions
```
Comment (zr2d2) [inline app/models/content.rb:212]: I guess you could do that down here and save a few lines, but i'm not too concerned either way

File: app/models/content.rb [hunk 9]
```diff
@@ -179,4 +181,39 @@ def perform_action(action_name, options)
...
+          filtered_contents.push(content)
+        end
+      end
+    else
+      # If filtering by user or type, we don't need to search through subscriptions, only contents
+      query_conditions[:user_id] = params[:user] if params[:user]
+      query_conditions[:type] = params[:type] if params[:type]
+      if query_conditions.empty?
+        # No filters are specified
+        filtered_contents = Content.all
```
Comment (zr2d2) [inline app/models/content.rb:209]: .find(:all) is depreciated, so i changed it to .all

**Pull request merged**